atc/scheduler/runner.go

```diff
@@ -34,7 +35,7 @@ type schedulerRunner struct {
 	running            *sync.Map
 }
 
-func NewRunner(logger lager.Logger, jobFactory db.JobFactory, scheduler BuildScheduler, maxJobs uint64) Runner {
+func NewRunner(logger lager.Logger, jobFactory db.JobFactory, scheduler BuildScheduler, maxJobs uint64) component.Runnable {
```

Comment:
If this returned `*schedulerRunner`, then we wouldn't need to depend on the component package here.

Reply:
👍  I can make the type public and just return it directly.

Reply:
You don't even have to make it public, private structs still conform to all the necessary interfaces and it stops other packages from using those types.

Reply:
Yeah, I just find that pattern kind of awkward. I've never really seen that in idiomatic Go codebases. 😕 Would be nice to just export the type and maybe even give it Godoc treatment. In cases like this I'd place more value on idiomatic than defensive code; we can catch that kind of stuff in PR review, and it seems unlikely to occur anyway.

Reply:
Why is it awkward? The whole point of having a constructor in the first place is because you want to control the creation of those objects.

Reply:
It's a pattern that seems to assume the codebase is a hostile environment with consumers of our APIs that don't know what they're doing. These seem like rare situations to catch in code review and tests.

We *could* abuse all kinds types exported by `stdlib` and third party packages by avoiding the constructor, but we don't. I've never really seen this pattern applied anywhere that we use.

It's awkward because it forces the caller to come up with an interface but only if they try to use the object in a certain way. Returning a private type means I can use the object so long as I do `foo := ...` and use `foo` within one function, but the moment I have to pass it as an argument somewhere (e.g. a private method) or if I use the `var foo` syntax, I can't be cause I can't refer to the type. So then the I have to explicitly declare an interface. Sure, it's nice having an explicit interface for the contract, but it really doesn't buy you anything if you weren't intending to have an interface already. The compiler will let you know if the interface breaks your expectations either way.

Returning a private type would also be a problem for the tests since they don't share the private namespace. They could just use the `component.Runnable` interface or declare their own, but then we've just moved the dependency from the package to the package's tests, or we've added an interface that we just have to keep changing to match the "real" interface.

To me, the upside of being more idiomatic code that can be documented with `godoc` outweighs the risk mitigated by having full control over the construction.

Reply:
> It's awkward because it forces the caller to come up with an interface but only if they try to use the object in a certain way. Returning a private type means I can use the object so long as I do foo := ... and use foo within one function, but the moment I have to pass it as an argument somewhere (e.g. a private method) or if I use the var foo syntax, I can't because I can't refer to the type. So then the I have to explicitly declare an interface. Sure, it's nice having an explicit interface for the contract, but it really doesn't buy you anything if you weren't intending to have an interface already.

You say this as if it's a bad thing lol. Forcing people to be explicit about what they accept by defining an interface is good design. It's also probably the most idiomatic thing about go.

Reply:
Honestly I don't really care about this specific case. I just want to make sure that we're being consistent with the rest of the codebase. And if we're not going to be consistent we need a good reason.

Reply:
I would prefer to be consistent with practices that are common in the Go community so that we can enforce them through tooling like `golint`, which now that I check it complains about this exact scenario:

```
atc/scheduler/runner.go:37:105: exported func NewRunner returns unexported type *scheduler.schedulerRunner, which can be annoying to use
```

Interfaces are great, but forcing someone to write an interface just for the sake of having one isn't intrinsically valuable. It can end up just being a maintenance burden and a mild headache as you can never 'go to definition' in your editor.